examples/light-switch-app/infineon/cyw30739/BUILD.gn

```diff
@@ -12,71 +12,54 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import("//build_overrides/build.gni")
 import("//build_overrides/chip.gni")
 import("//build_overrides/cyw30739_sdk.gni")
 
-import("${cyw30739_sdk_build_root}/cyw30739_executable.gni")
-import("${cyw30739_sdk_build_root}/cyw30739_sdk.gni")
+import("${chip_root}/examples/platform/infineon/cyw30739/args.gni")
+import("${matter_wpan_sdk_build_root}/matter_wpan_executable.gni")
 
+app_name = "light_switch_app"
 cyw30739_project_dir =
     "${chip_root}/examples/light-switch-app/infineon/cyw30739"
-examples_plat_dir = "${chip_root}/examples/platform/infineon/cyw30739"
 
-declare_args() {
-  setupPinCode = 20202021
-  setupDiscriminator = 3840
-}
-
-cyw30739_sdk("sdk") {
+matter_wpan_example("wpan_example") {
   sources = [ "${cyw30739_project_dir}/include/CHIPProjectConfig.h" ]
 
-  include_dirs = [
-    "${cyw30739_project_dir}/include",
-    "${examples_plat_dir}",
-  ]
-
-  defines = [
-    "CHIP_DEVICE_CONFIG_USE_TEST_SETUP_PIN_CODE=${setupPinCode}",
-    "CHIP_DEVICE_CONFIG_USE_TEST_SETUP_DISCRIMINATOR=${setupDiscriminator}",
-  ]
+  include_dirs = [ "${cyw30739_project_dir}/include" ]
 }
 
-cyw30739_executable("light_switch_app") {
-  output_name = "chip-cyw30739-light-switch-example.elf"
-
-  sources = [
-    "src/AppShellCommands.cpp",
-    "src/BindingHandler.cpp",
-    "src/ButtonHandler.cpp",
-    "src/LightSwitch.cpp",
-    "src/LightingManager.cpp",
-    "src/ZclCallbacks.cpp",
-    "src/main.cpp",
-  ]
+template("matter_wpan_app") {
+  forward_variables_from(invoker, [ "board" ])
 
-  deps = [
-    ":sdk",
-    "${chip_root}/examples/light-switch-app/light-switch-common",
-    "${chip_root}/examples/providers:device_info_provider",
-    "${chip_root}/examples/shell/shell_common:shell_common",
-    "${chip_root}/src/lib",
-  ]
+  import("${matter_wpan_sdk_build_root}/boards/${board}/args.gni")
 
-  include_dirs = [ "include" ]
+  matter_wpan_executable(target_name) {
+    sources = [
+      "src/AppShellCommands.cpp",
+      "src/AppTask.cpp",
+      "src/BindingHandler.cpp",
+      "src/ButtonHandler.cpp",
+      "src/LightSwitch.cpp",
+      "src/ZclCallbacks.cpp",
+    ]
 
-  if (chip_enable_ota_requestor) {
-    sources += [
-      "${examples_plat_dir}/OTAConfig.cpp",
-      "${examples_plat_dir}/OTAConfig.h",
+    deps = [
+      "${chip_root}/examples/shell/shell_common:shell_common",
+      app_data_model,
     ]
+
+    include_dirs = [ "include" ]
   }
 }
 
-group("cyw30739") {
-  deps = [ ":light_switch_app" ]
+foreach(board, matter_wpan_sdk_board_list) {
+  matter_wpan_app("${app_name}-${board}") {
```

Comment:
why do we build several apps in parallel? How about specifying the board itself as part of the build configuration?

Reply:
My goal is to minimize the increase in CI building time when I add more supported boards on the CYW30739 platform.
If we specified the board as part of the build configuration, the ninja would repeatedly build the whole Matter stack across every supported board.
But if we use the board as one of the GN targets, the GN would help us finger out the dependencies among app-specific codes, Matter stack codes, and board-specific codes. As a result, the ninja would only build board-independent parts of codes once, reducing the build time considerably.
In the CYW30739 case, building an app for a single board requires about 5 minutes, building an app for three different boards separately requires about 15 minutes, and building an app with three different boards simultaneously only requires about 7 minutes.

Reply:
I think long term this may not be workable - if you have more boards they would at some point differ under `/platform` in some way more than just peripherals seen by the app. This is why I would like top-level to keep board selection clear even if a single target can build multiple boards.